Issue w/ head.load

mattiaerre opened the pull request:
# Description

the following gif shows that when replacing `head.load` w/ `l` it is possible to wait until the asset has been loaded before firing a callback function; this PR includes also the 2 html files used in order to record the video.

the first part shows the behavior w/ `head.load` simulating a Regular 2G network w/ Chrome dev tool; the second part shows the behavior w/ `l` simulating the same network speed.

The test is fetching a `.css` and a `.js` file.

![ljs-2](https://user-images.githubusercontent.com/6423261/27344706-14b88384-559c-11e7-9e2e-5825552b64db.gif)

the [source code](https://github.com/malko/l.js) of the `l` library 

## Changes

- add the `l` library (and update .eslintignore)

- update `build.js` so that it is including `l` inside `oc-client.min.js` when `grunt build`

- use `l` instead of `head.load` inside karma-common (grunt karma:local still works)

mattiaerre commented:
this PR tries to fix #281 

// cc @matteofigus

mattiaerre commented on tasks/build.js:
```diff
@@ -7,46 +7,81 @@ const uglifyJs = require('uglify-js');
-      done();
-    });
-  });
+module.exports = function(grunt) {
```
this has been reformatted because of `prettier`

mattiaerre commented on src/components/oc-client/src/l.js:
```diff
@@ -0,0 +1,152 @@
+      }
+    }
+    //export ljs (as `head`)
+    window.head = loader;
```
this is the only `workaround` applied 😅

matteofigus replied:
Well this is temporary right? If we would want to go for it, we can change the library to make it use ljs instead and it should work, right?

mattiaerre replied:
do you want me to use `ljs` instead of `head` as the name of the loader? sure I'll try to do that. I will also test CSS loading.

matteofigus commented:
C'est formidable!

I think that this library does the job well, and despite not being super active, it makes all the tests pass, so I would say let's go for it.

I see that you did a good job on setting up a manual test too. The only think we may want to test (which I think is not covered by the karma tests) is css loading with oc.require. This library should support it according to the docs, but I just to ensure that we can keep all retro-compatible ;)

mattiaerre commented:
@matteofigus I updated this PR that now uses a global `ljs` name instead of `head`; I've also fixed the tests and tried to load a `.css` file (and re-recorded a short video w/ that). the only caveat w/ css is that the callback is raised synchronously before the load completion. But this happens for both `head.load` and `ljs`. It is how CSS loading works (I guess). If you are happy I can get rid of the examples and the `head.load` lib from this PR so that you can merge and build; unless you want me to include the build here.

mattiaerre commented on src/components/oc-client/src/l.js:
```diff
@@ -0,0 +1,152 @@
+      }
+    }
+    //export ljs
+    window.ljs = loader;
```
as per the lib now the global name is `ljs`

matteofigus commented:
Sounds good. Let's remember to update this file too https://github.com/opentable/oc/blob/master/src/components/oc-client/LICENSES

mattiaerre commented on src/components/oc-client/LICENSES:
```diff
@@ -27,7 +27,7 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 * jQuery
-* head.js
+* l.js
```
@matteofigus let me know if this is good for you

mattiaerre commented:
all 💚  over to you @matteofigus

matteofigus commented:
Fantastic. I decided to go live with a new minor (0.38.0) because despite this should be a non breaking change, there was a non explicit contract on exposing `head` so it is better to toggle a minor for clarity. For most of the user, this should be an easy upgrade.

mattiaerre commented:
I wanted to wait a couple of weeks (just to make sure that there was no production issue) but now I'd like to take some time and thank @malko for his `l.js` library that helped us smoothly to transition from `head.js`. Very good job and thanks again; OSS rulez 😎 

// cc @matteofigus @nickbalestra

malko commented:
@mattiaerre thank for your mention and use of ljs. Glad it can help you. As said earlier in the comment the repository is not really active but i want to assure you that this is not a dead project, I still accept PR and take a look to it on a regular basis. I just don't have more need for it at this time but as said PR are welcome as long as it don't forget the purpose of ljs, stay a micro library.
Hope you'll be happy with it.
